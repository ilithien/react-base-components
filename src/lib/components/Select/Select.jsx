import React, { Component } from "react";
import PropTypes from "prop-types";
import "./Select.css";

export default class Select extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      value: this.selectDefaultValue()
    };
  }

  selectDefaultValue = () => {
    const {
      placeholder,
      options: [first, ...rest],
      defaultValue
    } = this.props;
    if (defaultValue) {
      return defaultValue;
    } else if (placeholder) {
      return { label: placeholder };
    } else {
      return first;
    }
  };

  selectOption = option => {
    this.setState({ value: option, isOpen: false });
    this.props.onValueChange(option.value)
  };

  toggleOpen = () => {
    this.setState(prevState => ({
      isOpen: !prevState.isOpen
    }));
  };

  render() {
    const { options } = this.props;
    const {
      value: { label: currentLabel },
      isOpen
    } = this.state;

    return (
      <div className="react-base__select">
        <div className="react-base__select-selector" onClick={this.toggleOpen}>
          {currentLabel}
        </div>
        {isOpen && (
          <div className="react-base__select-options">
            {options.map(({ value, label }) => (
              <span
                key={value}
                onClick={() => this.selectOption({ value, label })}
              >
                {label}
              </span>
            ))}
          </div>
        )}
      </div>
    );
  }
}

Select.propTypes = {
  onValueChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      label: PropTypes.string
    }).isRequired
  ),
  defaultValue: PropTypes.string,
  placeholder: PropTypes.string
};
