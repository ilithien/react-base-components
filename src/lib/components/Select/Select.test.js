import React from "react";
import renderer from "react-test-renderer";
import Select from "./Select";

describe("Testing Select input component", () => {
  it("matches snapshot", () => {
    const snap = renderer.create(
      <Select
        options={[
          { value: 1, label: "Hey" },
          { value: 2, label: "You" },
          { value: 3, label: "Baby" }
        ]}
        onValueChange={() => console.log("Changing cool value bro")}
      />
    );

    const tree = snap.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
