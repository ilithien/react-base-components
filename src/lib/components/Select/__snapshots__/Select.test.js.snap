// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Select input component matches snapshot 1`] = `
<div
  className="react-base__select"
>
  <div
    className="react-base__select-selector"
    onClick={[Function]}
  >
    Hey
  </div>
</div>
`;
